### PR TITLE
classifier model that uses cortical.io's bitmap to represent a class

### PR DESCRIPTION
**Pull request by akhilaananthram**
At training time, represents each class with a bitmap.  At testing time, compares a sample to these bitmaps.

@BoltzmannBrain

### CODE REVIEW

**Comment by akhilaananthram** [comment]
@BoltzmannBrain: updated the model so it has the changes done to classification_model

File: fluent/models/classify_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    self.n = self.encoder.n
+    self.w = int((self.encoder.targetSparsity/100) * self.n)
+
+    self.positives = {}
+    self.categoryBitmaps = {}
+
+
+  def encodePattern(self, sample):
+    """
+    Encode an SDR of the input string by querying the Cortical.io API.
```
Comment (BoltzmannBrain) [inline fluent/models/classify_endpoint.py]: Update docstring

File: fluent/models/classify_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    text = " ".join(sample)
+    fpInfo = self.encoder.encode(text)
+    if self.verbosity > 1:
+      print "Fingerprint sparsity = {0}%.".format(fpInfo["sparsity"])
+    if fpInfo:
+      bitmap = numpy.array(fpInfo["fingerprint"]["positions"], dtype="uint32")
+      return {"text": text, "sparsity": fpInfo["sparsity"], "bitmap": bitmap}
+    else:
+      bitmap = numpy.empty(0)
+      return {"text": text, "sparsity": float(self.w)/self.n, "bitmap": bitmap}
```
Comment (BoltzmannBrain) [inline fluent/models/classify_endpoint.py]: Please revise this method to reflect the changes in #69; i.e. should be identical to `ClassificationModelFingerprint.encodePattern()`.

File: fluent/models/classify_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    self.positives[label].append(sample["text"])
+
+    categoryBitmap = self.client.createClassification(str(label), self.positives[label])["positions"]
+
+    self.categoryBitmaps[label] = categoryBitmap
+
+
+  def testModel(self, sample):
+    """
+    Test the kNN classifier on the input sample. Returns the classification most
```
Comment (BoltzmannBrain) [inline fluent/models/classify_endpoint.py]: Update docstring

File: fluent/models/classify_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    frequent amongst the classifications of the sample's individual tokens.
+    We ignore the terms that are unclassified, picking the most frequent
+    classification among those that are detected.
+
+    @param sample     (dictionary)      Dictionary, containing text, sparsity, and bitmap
+    @return           (dictionary)      The distances between the sample and the classes
```
Comment (BoltzmannBrain) [inline fluent/models/classify_endpoint.py]: Would be good to include an example showing what the keys are, [e.g.](https://github.com/numenta/nupic.fluent/blob/master/fluent/encoders/cio_encoder.py#L130-141)

File: fluent/models/classify_endpoint.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    @param label      (int)             Reference index for the classification
+                                        of this sample.
+    """
+    if label not in self.positives:
+      self.positives[label] = []
+    self.positives[label].append(sample["text"])
+
+    categoryBitmap = self.client.createClassification(str(label), self.positives[label])["positions"]
+
+    self.categoryBitmaps[label] = categoryBitmap
```
Comment (BoltzmannBrain) [inline fluent/models/classify_endpoint.py]: Why save the variable `categoryBitmap` at all?

File: fluent/models/classify_endpoint.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+  def resetModel(self):
+    """Reset the model"""
+    self.positives.clear()
+    self.categoryBitmaps.clear()
+
+
+  def trainModel(self, sample, label):
+    """
+    Train the classifier on the input sample and label.
```
Comment (BoltzmannBrain) [inline fluent/models/classify_endpoint.py]: Specify here (the unique behavior) that we are creating classification categories.

**Comment by BoltzmannBrain** [comment]
@akhilaananthram I left some inline notes. Also...
- `ClassificationModelEndpoint.trainModel()` should be able to include negative samples when creating categories, although we're not yet using this feature.
- Please be sure to always update docstrings to reflect changes.

Done with review.

**Comment by akhilaananthram** [comment]
@BoltzmannBrain: responded to feedback

**Comment by BoltzmannBrain** [comment]
:+1:

**Pull request merged**